File: INFORMED/A Search.c
#include <stdio.h>
#include <stdbool.h>
#include <limits.h>
#include <stdlib.h>
#include <time.h>

#define V 6

int minCost(int cost[], bool visited[])
{
    int min = INT_MAX, min_index;
    for (int v = 0; v < V; v++)
        if (visited[v] == false && cost[v] < min)
            min = cost[v], min_index = v;
    return min_index;
}

void printPath(int parent[], int j)
{
    if (parent[j] == -1)
        return;
    printPath(parent, parent[j]);
    printf(" -> %d ", j);
}

void printSolution(int cost[], int parent[], int src, int dest)
{
    printf("Path from %d to %d: %d", src, dest, src);
    printPath(parent, dest);
    printf("\nTotal Cost: %d\n", cost[dest]);
}

void aStarSearch(int graph[V][V], int src, int dest, int heuristic[])
{
    int parent[V];
    bool visited[V];
    int cost[V];
    for (int i = 0; i < V; i++)
    {
        parent[src] = -1;
        visited[i] = false;
        cost[i] = INT_MAX;
    }
    cost[src] = 0;
    for (int count = 0; count < V - 1; count++)
    {
        int u = minCost(cost, visited);
        visited[u] = true;
        for (int v = 0; v < V; v++)
        {
            if (!visited[v] && graph[u][v] && cost[u] + graph[u][v] + heuristic[v] < cost[v])
            {
                parent[v] = u;
                cost[v] = cost[u] + graph[u][v] + heuristic[v];
            }
        }
    }
    printSolution(cost, parent, src, dest);
}

int main()
{
    int graph[V][V];
    printf("Masukkan biaya perjalanan antar-stasiun:\n");
    int inputCost;
    scanf("%d", &inputCost);

    srand(time(NULL));

    for (int i = 0; i < V; i++)
    {
        for (int j = 0; j < V; j++)
        {
            if (i != j)
            {
                graph[i][j] = rand() % inputCost + 1;
            }
            else
            {
                graph[i][j] = 0;
            }
        }
    }

    int heuristic[V];
    printf("Nilai heuristik untuk setiap stasiun (acak):\n");
    for (int i = 0; i < V; i++)
    {
        heuristic[i] = rand() % 10;
        printf("Heuristik pada stasiun %d: %d\n", i, heuristic[i]);
    }

    int src, dest;
    printf("Masukkan stasiun awal (nomor): ");
    if (scanf("%d", &src) != 1 || src < 0 || src >= V)
    {
        printf("Input tidak valid.\n");
        return 1;
    }
    printf("Masukkan stasiun tujuan (nomor): ");
    if (scanf("%d", &dest) != 1 || dest < 0 || dest >= V)
    {
        printf("Input tidak valid.\n");
        return 1;
    }
    aStarSearch(graph, src, dest, heuristic);
    return 0;
}

File: INFORMED/ASearch.py
import time

# in case merupakan jumlah stasiun yang ada
V = 6

def min_cost(cost, visited):

    # initialisasi biaya minimum
    min_val = float('inf') 
    # initialisasi index minimum
    min_index = -1
    for v in range(V):
        if not visited[v] and cost[v] < min_val:
            min_val = cost[v]
            min_index = v
    return min_index

def print_path(parent, j):
    if parent[j] == -1:
        return
    # rekursif mencetak jalur
    print_path(parent, parent[j])
    print(" ->", j, end='')

def print_solution(cost, parent, src, dest):
    # fungsi untuk mencetak jalur yang ditemukan dari stasiun awal ke stasiun tujuan, beserta total biaya perjalanan.
    print("Path from", src, "to", dest, ":", src, end='')
    print_path(parent, dest)
    print("\nTotal Cost:", cost[dest])

def a_star_search(graph, src, dest, heuristic):
    # array untuk menyimpan orang tua dari setiap stasiun yang dikunjungi.
    parent = [-1] * V
    # array untuk menandai stasiun yang telah dikunjungi.
    visited = [False] * V
    # array untuk menyimpan biaya terpendek dari stasiun awal ke setiap stasiun.
    cost = [float('inf')] * V

    # karan source awal adalah stasiun awal, jadi tidak ada biaya
    cost[src] = 0

    for _ in range(V - 1):
        u = min_cost(cost, visited)
        visited[u] = True
        for v in range(V):
            if not visited[v] and graph[u][v] and cost[u] + graph[u][v] + heuristic[v] < cost[v]:
                parent[v] = u
                cost[v] = cost[u] + graph[u][v] + heuristic[v]

    print_solution(cost, parent, src, dest)

def main():
    graph = [[0] * V for _ in range(V)]
    print("Masukkan biaya perjalanan antar-stasiun:")
    input_cost = int(input())

    for i in range(V):
        for j in range(V):
            if i != j:
                graph[i][j] = (i + j + 1) % input_cost + 1
            else:
                graph[i][j] = 0

    heuristic = []
    print("Nilai heuristik untuk setiap stasiun (acak):")
    for i in range(V):
        heuristic.append((i * 3) % 10)
        print("Heuristik pada stasiun", i, ":", heuristic[i])

    src = int(input("Masukkan stasiun awal (nomor): "))
    if src < 0 or src >= V:
        print("Input tidak valid.")
        return

    dest = int(input("Masukkan stasiun tujuan (nomor): "))
    if dest < 0 or dest >= V:
        print("Input tidak valid.")
        return

    a_star_search(graph, src, dest, heuristic)

if __name__ == "__main__":
    main()
